### PR TITLE
custom validation for non-patientbased qdm population criteria

### PR DESCRIPTION
**Pull request by sb-cecilialiu**
## MADiE PR

Jira Ticket: [MAT-5537](https://jira.cms.gov/browse/MAT-5537)
(Optional) Related Tickets:

### All Submissions
* [x] This PR has the JIRA linked.
* [x] Required tests are included.
* [x] No extemporaneous files are included (i.e Complied files or testing results).
* [x] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [ ] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [x] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).
* [ ] All CDN/Web dependencies are hosted internally (i.e MADiE-Root Repo).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx [hunk 1]
```diff
@@ -449,6 +451,53 @@ const MeasureGroups = (props: MeasureGroupProps) => {
...
+            firstString = firstString.toLowerCase();
+          }
+          const secondString = definition.substring(
+            definition.indexOf(" ") + 1
+          );
+          if (secondString) {
+            secondString.replace(" ", "");
+          }
+          newDefintion = firstString + secondString;
+        }
```
Comment (adongare) [inline src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx]: seems like this whole piece is equivalent to loadash `_.camelCase`

File: src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx [hunk 2]
```diff
@@ -449,6 +451,53 @@ const MeasureGroups = (props: MeasureGroupProps) => {
...
   };
 
+  const validatePopulations = (
+    populations: Population[],
+    measureObservations: MeasureObservation[]
+  ): string => {
+    const returnTypesSet: Set<String> = new Set();
+    populations.forEach((population) => {
+      if (population.definition != undefined) {
+        const name = population.name;
```
Comment (adongare) [inline src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx]: this is unused?

File: src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx [hunk 3]
```diff
@@ -449,6 +451,53 @@ const MeasureGroups = (props: MeasureGroupProps) => {
...
+    });
+    if (measureObservations && measureObservations.length > 0) {
+      measureObservations.forEach((observation) => {
+        const returnType =
+          cqlDefinitionDataTypes[_.camelCase(observation.definition)];
+        if (returnType !== undefined) {
+          returnTypesSet.add(returnType);
+        }
+      });
+    }
```
Comment (adongare) [inline src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx]: looks like we need to get return types multiple times for populations, observations, and maybe for stratifications in the future. we could write a general function to get the return types for selected defines?
e.g.
```
const getReturnTypes = (populations: Array<Population | MeasureObservation>) => {
    return populations.reduce((returnTypes, population) => {
      const definition = _.camelCase(_.trim(population?.definition));
      const returnType = cqlDefinitionDataTypes[definition];
      if (returnType) {
        returnTypes.push(returnType);
      }
      return returnTypes;
    }, []);
  };
```
which then can be called 

```
    const populationReturnTypes = getReturnTypes(populations);
    const observationReturnTypes = getReturnTypes(measureObservations);
    const returnTypes = new Set([
      ...populationReturnTypes,
      ...observationReturnTypes,
    ]);
```

File: src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx (hunk 1 above)
Comment (sb-cecilialiu) [reply src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx]: @adongare Didn't know that, thanks so much Ashok!

File: src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx (hunk 2 above)
Comment (sb-cecilialiu) [reply src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx]: @adongare No it wasn't.

File: src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx (hunk 3 above)
Comment (sb-cecilialiu) [reply src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx]: @adongare This is great! A lot neater thanks so much Ashok! Just changed to use it.

Comment (adongare) [reply src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx]: Sure, np

File: src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx [hunk 4]
```diff
@@ -453,6 +456,56 @@ const MeasureGroups = (props: MeasureGroupProps) => {
...
+      const returnType = cqlDefinitionDataTypes[definition];
+      if (returnType) {
+        returnTypes.push(returnType);
+      }
+      return returnTypes;
+    }, []);
+  };
+  const getReturnTypesStrats = (populations: Array<Stratification>) => {
+    return populations?.reduce((returnTypes, population) => {
+      const definition = _.camelCase(_.trim(population?.cqlDefinition));
```
Comment (adongare) [inline src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx:473]: I wish it was `definition` instead of `cqlDefinition`

Comment (sb-cecilialiu) [reply src/components/editMeasure/populationCriteria/groups/QDM/QDMMeasureGroups.tsx:473]: @adongare Yeah i was thinking the same!!

**Review by adongare — APPROVED** [review]

**Pull request merged**